File: ts/src/relations.ts
import { z } from "zod";
import { RefError } from "./errors.js";

export const RELATES_TO = "relatesTo";

export const RelationSchema = z.object({
  source: z.string(),
  predicate: z.string(),
  target: z.string(),
  directed: z.boolean().default(true),
  weight: z.number().nullable().default(null),
  attrs: z.record(z.unknown()).default({}),
});

export type Relation = z.infer<typeof RelationSchema>;

// Low-level schema helpers. Like Python's `Relation(...)` / `Relation.from_serialized`, these may
// surface a raw ZodError on malformed input. The "never leak raw" contract lives in the boundary
// parsers (`makeNode`, `nodeFromMarkdown`, `membershipOf`), which catch and re-wrap as kernel errors.
export function fromSerialized(data: Record<string, unknown>, containerId: string): Relation {
  const { source, ...rest } = data;
  return RelationSchema.parse({ source: source !== undefined ? source : containerId, ...rest });
}

export function toSerialized(rel: Relation, containerId: string): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (rel.source !== containerId) out.source = rel.source;
  out.predicate = rel.predicate;
  out.target = rel.target;
  if (rel.directed !== true) out.directed = rel.directed;
  if (rel.weight !== null) out.weight = rel.weight;
  if (Object.keys(rel.attrs).length > 0) out.attrs = rel.attrs;
  return out;
}

export function relatesTo(source: string, target: string): Relation {
  return RelationSchema.parse({ source, predicate: RELATES_TO, target });
}

export function tagToRelation(source: string, tag: string, aliasMap: Record<string, string>): Relation {
  const name = tag.replace(/^#+/, "");
  const target = aliasMap[name] ?? aliasMap[name.toLowerCase()];
  if (target === undefined) {
    throw new RefError(`tag ${JSON.stringify(tag)} does not resolve to a known node`);
  }
  return relatesTo(source, target);
}

File: python/src/nodes/core/relations.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from nodes.core.errors import RefError

RELATES_TO = "relatesTo"


class Relation(BaseModel):
    """The single edge primitive (normalized form: source always explicit)."""

    source: str
    predicate: str
    target: str
    directed: bool = True
    weight: float | None = None
    attrs: dict[str, Any] = Field(default_factory=dict)

    @classmethod
    def from_serialized(cls, data: dict, container_id: str) -> "Relation":
        return cls(
            source=data.get("source", container_id),
            predicate=data["predicate"],
            target=data["target"],
            directed=data.get("directed", True),
            weight=data.get("weight"),
            attrs=data.get("attrs", {}),
        )

    def to_serialized(self, container_id: str) -> dict:
        out: dict[str, Any] = {}
        if self.source != container_id:
            out["source"] = self.source
        out["predicate"] = self.predicate
        out["target"] = self.target
        if self.directed is not True:
            out["directed"] = self.directed
        if self.weight is not None:
            out["weight"] = self.weight
        if self.attrs:
            out["attrs"] = self.attrs
        return out


def relates_to(source: str, target: str) -> Relation:
    return Relation(source=source, predicate=RELATES_TO, target=target)


def tag_to_relation(source: str, tag: str, alias_map: dict[str, str]) -> Relation:
    name = tag.lstrip("#")
    target = alias_map.get(name) or alias_map.get(name.lower())
    if target is None:
        raise RefError(f"tag {tag!r} does not resolve to a known node")
    return relates_to(source, target)
